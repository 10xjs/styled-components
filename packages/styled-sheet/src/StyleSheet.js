// @flow

import type { GroupedTag, GroupedKeys } from './types';
import { IS_BROWSER } from './constants';
import { RuleGroupTag, DefaultTag } from './tags';
import { makeCssMarker } from './utils';
import GroupRegistry from './GroupRegistry';
import rehydrate from './rehydrate';

const hasRehydrated = !IS_BROWSER;

class StyleSheet {
  groups: GroupedTag;

  keys: GroupedKeys;

  constructor(target?: HTMLElement) {
    // $FlowFixMe
    this.keys = Object.create(null);

    const tag = new DefaultTag(target);
    this.groups = new RuleGroupTag(tag);

    if (!hasRehydrated) {
      this.rehydrate();
    }
  }

  rehydrate() {
    rehydrate(this.groups);
  }

  register(name: string): number {
    return GroupRegistry.registerRuleGroup(name);
  }

  hasKey(group: number, key: string): boolean {
    const groupKeys = this.keys[group];
    return groupKeys !== undefined && groupKeys[key] !== undefined;
  }

  registerKey(group: number, key: string) {
    let groupKeys = this.keys[group]
    if (groupKeys === undefined) {
      // $FlowFixMe
      groupKeys = (this.keys[group] = Object.create(null));
    }

    groupKeys[key] = true;
  }

  inject(group: number, key: string, rules: string[]) {
    if (!this.hasKey(group, key)) {
      this.registerKey(group, key);
      this.groups.insertRules(group, rules);
    }
  }

  remove(group: number) {
    // $FlowFixMe
    this.keys[group] = Object.create(null);
    this.groups.clearGroup(group);
  }

  toString(): string {
    let css = '';
    GroupRegistry.forEach((name, group) => {
      const groupRules = this.groups.getGroup(group);
      if (groupRules !== '') {
        const keysForGroup = Object.keys(this.keys[group] || {});
        const marker = makeCssMarker(name, group, keysForGroup);
        css += `${marker}\n${groupRules}`;
      }
    });

    return css;
  }
}

export default StyleSheet;
