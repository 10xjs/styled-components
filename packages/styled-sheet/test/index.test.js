import StyleSheet, { DOMStyleSheet } from '../src';

it('should insert styles', () => {
  const sheet = new DOMStyleSheet();
  const styleSheet = new StyleSheet(sheet);
  styleSheet.insertRules(['.bla { color: red; }'], 0);
  styleSheet.insertRules(['.bla { color: red; }'], 0);
  styleSheet.insertRules(['.bla2 { color: red; }'], 0);
  styleSheet.insertRules(['.bla2 { color: red; }'], 0);
  expect(sheet.tag.sheet.cssRules).toMatchSnapshot();
});
