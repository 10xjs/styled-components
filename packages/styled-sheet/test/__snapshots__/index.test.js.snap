// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should insert styles 1`] = `
Array [
  CSSStyleRule {
    "__ends": 20,
    "__starts": 0,
    "parentRule": null,
    "parentStyleSheet": CSSStyleSheet {
      "cssRules": [Circular],
      "parentStyleSheet": null,
    },
    "selectorText": ".bla",
    "style": CSSStyleDeclaration {
      "0": "color",
      "__starts": 5,
      "_importants": Object {
        "color": "",
      },
      "color": "red",
      "length": 1,
      "parentRule": [Circular],
    },
  },
  CSSStyleRule {
    "__ends": 20,
    "__starts": 0,
    "parentRule": null,
    "parentStyleSheet": CSSStyleSheet {
      "cssRules": [Circular],
      "parentStyleSheet": null,
    },
    "selectorText": ".bla",
    "style": CSSStyleDeclaration {
      "0": "color",
      "__starts": 5,
      "_importants": Object {
        "color": "",
      },
      "color": "red",
      "length": 1,
      "parentRule": [Circular],
    },
  },
  CSSStyleRule {
    "__ends": 21,
    "__starts": 0,
    "parentRule": null,
    "parentStyleSheet": CSSStyleSheet {
      "cssRules": [Circular],
      "parentStyleSheet": null,
    },
    "selectorText": ".bla2",
    "style": CSSStyleDeclaration {
      "0": "color",
      "__starts": 6,
      "_importants": Object {
        "color": "",
      },
      "color": "red",
      "length": 1,
      "parentRule": [Circular],
    },
  },
  CSSStyleRule {
    "__ends": 21,
    "__starts": 0,
    "parentRule": null,
    "parentStyleSheet": CSSStyleSheet {
      "cssRules": [Circular],
      "parentStyleSheet": null,
    },
    "selectorText": ".bla2",
    "style": CSSStyleDeclaration {
      "0": "color",
      "__starts": 6,
      "_importants": Object {
        "color": "",
      },
      "color": "red",
      "length": 1,
      "parentRule": [Circular],
    },
  },
]
`;
